### ISSUE TITLE
keyof widen to `string | number` for  `Record<string, any>` and `{[id: string]: any}`

### ISSUE THREAD

**Issue by sirian** [Microsoft/TypeScript#30883]
**TypeScript Version:**  3.4.3, 3.5.0-dev.20190412

**Code**

```ts
declare function get<T, K extends keyof T>(t: T, k: K): T[K];

declare const x: {[id: string]: boolean};
const x1: boolean = x[3];
const x2: boolean = get(x, 3);
const xk: keyof typeof x = 3;

declare const y: Record<string, boolean>;
const y1: boolean = y[3];
const y2: boolean = get(y, 3); // ERROR
const yk: keyof typeof x = 3;

const xy: typeof x = y;
const yx: typeof y = x; 
```

**Expected behavior:**
Inconsistency between `y2` and `y1`, `yk` (and `x1`,`x2`,`xk`). 
IMO `keyof x` and `keyof y` should be `string`, not `number | string`. Or at least all examples should be consistent (either fails or not)

**Actual behavior:**
Error shown only for y2 (line 10). But at the same time no error shown for `y1` and `yk`. 

![image](https://user-images.githubusercontent.com/897643/56034919-69200e80-5d31-11e9-9b2a-0adc4cf65b6d.png)

**Playground Link:** [playground](https://www.typescriptlang.org/play/index.html#src=declare%20function%20get%3CT%2C%20K%20extends%20keyof%20T%3E(t%3A%20T%2C%20k%3A%20K)%3A%20T%5BK%5D%3B%0A%0Adeclare%20const%20x%3A%20%7B%5Bid%3A%20string%5D%3A%20boolean%7D%3B%0Aconst%20x1%3A%20boolean%20%3D%20x%5B3%5D%3B%0Aconst%20x2%3A%20boolean%20%3D%20get(x%2C%203)%3B%0Aconst%20xk%3A%20keyof%20typeof%20x%20%3D%203%3B%0A%0Adeclare%20const%20y%3A%20Record%3Cstring%2C%20boolean%3E%3B%0Aconst%20y1%3A%20boolean%20%3D%20y%5B3%5D%3B%0Aconst%20y2%3A%20boolean%20%3D%20get(y%2C%203)%3B%20%2F%2F%20ERROR%0Aconst%20yk%3A%20keyof%20typeof%20x%20%3D%203%3B%0A%0Aconst%20xy%3A%20typeof%20x%20%3D%20y%3B%0Aconst%20yx%3A%20typeof%20y%20%3D%20x%3B%20)

**Related Issues:** 
https://github.com/Microsoft/TypeScript/issues/29249

**Comment by RyanCavanaugh** [comment]
This was intentional. Pretty much any decision here introduces inconsistency; see see #13715 / #23592 and the many related thread there.

You can declare
```ts
declare const y: Record<string | number, boolean>;
```
if you'd prefer the other behavior.

**Comment by sirian** [comment]
@RyanCavanaugh 
Sorry, I made a typo at line 11. yk should be `typeof keyof Record<string, any>` but due to type it was `typeof keyof {[id: string]: any}` and was inferred as `number | string` due to a typo. And this confused me (inconsitence between line 10 and 11). 

 Looks like it's ok after fixing typo


![image](https://user-images.githubusercontent.com/897643/56080938-ede15a00-5e0f-11e9-8586-1430d16bdce0.png)
